### PR TITLE
Issue#26

### PR DESCRIPTION
**Pull request by siddhero97**
### What problem are you solving in this PR?
The user should be redirected to the homepage if he or she isn't logged in,

### Related issues/PRs:
resolve https://github.com/siddhero97/CS436I-Memofy/issues/26

### How is this accomplished and what led to this approach?
Whenever the user logins it, an action called logIn is called which updates the currUser and changes isLogged in within UserState to true. If no user is logged in then the isLoggedIn is false. I have assumed that there is already a set of registered users, from which to work on.  

### What are the risks in this PR?
I could potentially mess up the functionality of the login page if not done right. 

### Before-I-merge checklist
- [] I have [self-tophatted]

### CODE REVIEW

**Comment by jarhun88** [comment]
why is Users an array? shouldnt there only ever be one user logged in at a time per session?
also how can i check that the redux is working? the redux devtool is just showing undefined for me

**Comment by vincentchiang1** [comment]
i believe we talked about having more user fields like email, phone, account type (ie. premium) in our last meeting

**Comment by siddhero97** [comment]
> why is Users an array? shouldn't there only ever be one user logged in at a time per session?
> also how can i check that the redux is working? the redux devtool is just showing undefined for me

1)The users represent the registeredUsers. I imagine that since we don't have a registration button. There will be a set of registered users somewhere. Whenever a user logs in, we should check whether the user is included in a set of registered users. 
2) I have posted 2 videos with them working.

**Comment by siddhero97** [comment]
> i believe we talked about having more user fields like email, phone, account type (ie. premium) in our last meeting

I'm trying to get a basic idea about whether what I am doing is working and whether its ok. Adding more fields shouldn't be 2 hard.

**Comment by KwanLam17** [comment]
> it'd be nice to mention what the ticket is about in the title or even a brief description instead of just saying the issue #. also why are we using the user story template if it's completely blank.

This is a very good suggestion, something that I have been itching to say. Please name your branches with more meaning too. This is the convention I follow for naming branches:

`[issue#]-[description]`

The description part could have dashes, so an example for this issue would be:

`26-add-user-data`

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 1]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
 import {Button, ButtonGroup} from '@shopify/polaris';
 
 import './Navbar.css';
+import {useDispatch} from "react-redux";
+import {logout} from "../../../../store/users/action";
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:7]: I added aliases to the base folders, avoid having more than 2 directory "up"s.

`import {logout} from 'store/users/action;`

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 2]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
+import {initialState} from "../../../../store/users/reducers";
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:8]: Same thing here with the aliasing. Also, you should not be accessing the store directly like this. If you want the initial state, get it from the redux store directly. You did not implement selectors anywhere, please look at my changes to see what I did.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 3]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
 
 export default function Navbar() {
   const history = useHistory();
+  const dispatch = useDispatch();
 
+  const redirection = () => {
+    // console.log("Inside Redirection");
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:15]: Remove commented code.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 4]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
+    if(!initialState.isLoggedIn) {
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:16]: Why is the Navbar in charge of checking if the user is logged in? This should be done in the `<Fridge />` component.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 5]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
   const handleHomeClick = () => {
     history.push('./home');
   };
 
   const handleLogoutClick = () => {
     history.push('./login');
+    dispatch(logout());
   };
-
+  window.onload = redirection;
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:30]: By checking if the user is logged in the `<Fridge />` component you don't need this direct access to the window DOM element.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 6]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
+    }
+    return null;
+  };
   const handleHomeClick = () => {
     history.push('./home');
   };
 
   const handleLogoutClick = () => {
     history.push('./login');
+    dispatch(logout());
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:28]: You should dispatch the action first before pushing the history. In fact, if the `isLoggedIn` is checked in the `<Fridge />` component you probably don't even need to push to history.

**Comment by siddhero97** [comment]
> > it'd be nice to mention what the ticket is about in the title or even a brief description instead of just saying the issue #. also why are we using the user story template if it's completely blank.
> 
> This is a very good suggestion, something that I have been itching to say. Please name your branches with more meaning too. This is the convention I follow for naming branches:
> 
> `[issue#]-[description]`
> 
> The description part could have dashes, so an example for this issue would be:
> 
> `26-add-user-data`

I will remember that for my next commit.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 7]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
+  const redirection = () => {
+    // console.log("Inside Redirection");
+    if(!initialState.isLoggedIn) {
+      alert("You Are Not Logged In!!!");
+      history.push('./login');
+    }
+    return null;
+  };
   const handleHomeClick = () => {
     history.push('./home');
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:23]: I made this mistake when I made my PR, can you get rid of the dot so it's just `/home`?

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 8]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
   };
 
   const handleLogoutClick = () => {
     history.push('./login');
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:27]: I made this mistake when I made my PR, can you get rid of the dot so it's just `/login`?

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx [hunk 9]
```diff
@@ -3,19 +3,33 @@ import {useHistory} from 'react-router-dom';
...
+    dispatch(logout());
   };
-
+  window.onload = redirection;
   return (
+
```
Comment (KwanLam17) [inline client/src/sections/Fridge/Header/Navbar/Navbar.tsx:32]: Remove empty line.

File: client/src/sections/Login/Login.tsx [hunk 10]
```diff
@@ -4,12 +4,18 @@ import {Button, Card, Heading, ButtonGroup} from '@shopify/polaris';
 import logo from 'icons/logo.jpg';
 
 import './Login.css';
+import {useDispatch} from "react-redux";
+import {login} from "../../store/users/action";
```
Comment (KwanLam17) [inline client/src/sections/Login/Login.tsx:8]: Aliases.

File: client/src/sections/Login/Login.tsx [hunk 11]
```diff
@@ -4,12 +4,18 @@ import {Button, Card, Heading, ButtonGroup} from '@shopify/polaris';
...
+import {User} from "../../store/users/types";
```
Comment (KwanLam17) [inline client/src/sections/Login/Login.tsx:9]: Aliases.

File: client/src/store/types.ts [hunk 12]
```diff
@@ -1,9 +1,15 @@
...
 
 export type RootState = {
   item: ItemState;
 };
 
 export interface ItemState {
   items: Item[];
+}
+export interface UserState {
+  users: User[];
```
Comment (KwanLam17) [inline client/src/store/types.ts:12]: I see @jarhun88 started this discussion about why we have an array of Users. I agree, this should only be one user. You don't need to make a User type at all at this point.

You mentioned that our app will support more than one user. That is correct, but that is not the responsibility of redux or "states" in general. The (redux) state of the application is supposed to only tell us what the state is for that particular web browser, it is not to represent the entire landscape of the app (in this case, all the users of our application).

Imagine if we had 100k users, do we want to store 100k users on the client side? Is that necessary?

We will solve the problem of managing multiple users once we have a database set up and we utilize third party logins to identify users.

File: client/src/store/types.ts [hunk 13]
```diff
@@ -1,9 +1,15 @@
...
+  isLoggedIn: boolean;
```
Comment (KwanLam17) [inline client/src/store/types.ts:13]: This boolean is the responsibility of the App state. The user state just describe the user who is currently logged in, and used for fetching data about them when they load different pages. The app state determines if someone is logged in, regardless of who the user is.

File: client/src/store/users/reducers.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+];
+
+export const initialState: UserState = {
+  users: initialUsers,
+  isLoggedIn: false,
+  currUser: null,
+};
+function isLoggedIn(users: User[]): boolean {
+  let isLoggedIn = false;
+  for(const user of users) {
```
Comment (KwanLam17) [inline client/src/store/users/reducers.ts:24]: This is a perfect example of why we should not have an array of users in the redux store. We probably don't want to loop through 100k users on the client side. We can remove this reducer.

File: client/src/store/users/types.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
+export interface User {
+    name: string;
+    userName: string;
+    isLoggedIn: boolean;
+}
+
+export const LOGIN = 'LOGIN';
+export const IS_LOGIN = 'IS_LOGIN';
```
Comment (KwanLam17) [inline client/src/store/users/types.ts:8]: Remove, this is not being used and will not be used here.

File: client/src/store/users/reducers.ts [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    if(user.isLoggedIn) {
+      isLoggedIn = true;
+      return  isLoggedIn;
+    }
+  }
+  return  isLoggedIn;
+}
+export function userReducer(
```
Comment (KwanLam17) [inline client/src/store/users/reducers.ts:32]: This reducer needs to be re-implemented.

**Comment by KwanLam17** [comment]
> it'd be nice to mention what the ticket is about in the title or even a brief description instead of just saying the issue #. also why are we using the user story template if it's completely blank.

@vincentchiang1 what do you mean is completely blank?

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx (hunk 6 above)
Comment (siddhero97) [reply client/src/sections/Fridge/Header/Navbar/Navbar.tsx:28]: The way that I thought about is that when somebody presses logout then shouldn't we logout. I will dispatch before pushing to history.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx (hunk 7 above)
Comment (siddhero97) [reply client/src/sections/Fridge/Header/Navbar/Navbar.tsx:23]: Sure, np.

File: client/src/sections/Fridge/Header/Navbar/Navbar.tsx (hunk 8 above)
Comment (siddhero97) [reply client/src/sections/Fridge/Header/Navbar/Navbar.tsx:27]: Sure, np

**Pull request merged**